HIP/hip-679.md

```diff
@@ -0,0 +1,215 @@
+    - If no results, attempt to list signatures from a path with node IDs
+    - If successful, set `pathType` as `NODE_ID` so the node ID path is used going forward
+   
+In the existing scenario the default pathType would be `ACCOUNT_ID`, which would then be changed to AUTO when we are close to approaching Phase two.
+
+#### Phase Two:
+Update the consensus nodes to write stream files to the new `{network}/{shard}/{nodeID}/*` nodeId based path.
+
+#### Phase Three:
+Create a new lambda function which performs data rollup along with compression.
+
+##### Compression zstd vs gz:
+- zstd has a better compression ration that gzip.
+- (zstd:gzip)
+  - Faster compression - About a 5:1  ratio.
+  - Faster decompression - About a 3:1 ratio.
+
+##### Algorithm:
+- Read stream files from current location 
+    ``` 
+        accountBalances/balance0.0.3/ 
+        eventsStreams/events_0.0.3/     
+        recordstreams/record0.0.3/ 
+    ```
+    
+- The new path for aggregated files would be:
+    
+    ``` 
+        {network}/{shard}/record/
+        {network}/{shard}/balance/
+        {network}/{shard}/event/
+    ```
+- The lambda will decompress the file, if it is compressed.
+- Calculate a hash of the file contents.
+- Maintain an in-memory map which has the hashes for the current file being processed, to remove duplicate files with the same content.
+- Compress the file using zst if it not a signature file.
+- Monitor the size/time of the potential aggregated file.
+- For aggregation:
+  - Calculate the node Id based on the current node account id. NodeId = (number portion of) nodeAccountId - 3
+  ``` Node for balance0.0.4 would be 4-3 = 1  ```
+  - The new file structure would be as follows:
+    Example:
+    ```
+    >tar -tf 2022-07-13T08_46.tar | sort  
+
+    2022-07-13T08_46/
+    2022-07-13T08_46/0/
+    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd.zst
+    2022-07-13T08_46/0/2022-07-13T08_46_08.041986003Z.rcd_sig
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd.zst
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z.rcd_sig
+    2022-07-13T08_46/0/2022-07-13T08_46_11.304284003Z_01.rcd.zst
+    2022-07-13T08_46/1/
+    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd.zst
+    2022-07-13T08_46/1/2022-07-13T08_46_08.041986003Z.rcd_sig
+    
+     ```
+     The above compressed tar file includes the following files in node id based folders:
+     - Record stream files (.rcd.zst)
+     - Signature files (.rcd_sig)
+     - Sidecar files (_01.rcd.zst)
+     
+     NOTE: File `2022-07-13T08_46_08.041986003Z.rcd.zst` appears in two separate node folders. These are files with distinct content and when downloaded by the mirror node one of the two file will be selected based on the consensus majority.
+     
+     The aggregation in the new files would be size/time based, as will be determined by performance testing.
```

Comment:
Are we going to try multi levels of roll up? Hour/Day/Week/Month/Year etc?

Reply:
Just trying to understanding how that would be helpful? Currently the idea is to not do that , but as you see the rollup name is to the minute.

Reply:
If we are running an importer that is trying to catch up, I found I was performance limited by number of API calls per second to storage bucket. There is one per file, so less files = faster. My thought that having a roll up of a complete days data in a single compressed file would super speed up bulk import.

Reply:
As mentioned in the HIP, as part of engineering design and with the help of performance testing , we still need to determine what the optimum size of the rollup would be.